src/main/cpp/src/SparkResourceAdaptorJni.cpp

```diff
@@ -52,121 +225,832 @@ class spark_resource_adaptor final : public rmm::mr::device_memory_resource {
+      case SHUFFLE_ALLOC_FREE:
+          transition(thread->second, thread_state::SHUFFLE_RUNNING);
+          break;
+      default:
+          break;
+      }
+      wake_next_highest_priority_regular_blocked(lock);
+    }
+  }
+
+  /**
+   * Wake the highest priority blocked (not BUFN) thread so it can make progress.
+   * This is typically called when a free happens, or an alloc succeeds.
+   */
+  void wake_next_highest_priority_regular_blocked(std::unique_lock<std::mutex> & lock) {
+    // 1. Find the highest priority blocked thread, including shuffle.
+    thread_priority to_wake(-1, -1);
+    bool is_to_wake_set = false;
+    for (auto thread = threads.begin(); thread != threads.end(); thread++) {
+      thread_state state = thread->second.state;
+      if (state == thread_state::TASK_BLOCKED || state == thread_state::SHUFFLE_BLOCKED) {
+        thread_priority current = thread->second.priority();
+        if (!is_to_wake_set || to_wake < current) {
+          to_wake = current;
+          is_to_wake_set = true;
+        }
+      }
+    }
+    // 2. wake up that thread
+    long thread_id_to_wake = to_wake.get_thread_id();
+    if (thread_id_to_wake > 0) {
+      auto thread = threads.find(thread_id_to_wake);
       if (thread != threads.end()) {
-        if (thread->second.retry_oom_injected > 0) {
-          thread->second.retry_oom_injected--;
-          throw_java_exception(RETRY_OOM_CLASS, "injected RetryOOM");
+        switch (thread->second.state) {
+          case TASK_BLOCKED:
+            transition(thread->second, thread_state::TASK_RUNNING);
+            task_wake_condition.notify_all();
+            break;
+          case SHUFFLE_BLOCKED:
+            transition(thread->second, thread_state::SHUFFLE_RUNNING);
+            task_wake_condition.notify_all();
+            break;
+          default:
+            {
+              std::stringstream ss;
+              ss << "internal error expected to only wake up blocked threads " << 
+                thread_id_to_wake << " " << as_str(thread->second.state);
+              throw std::runtime_error(ss.str());
+            }
         }
+      }
+    }
+  }
 
-        if (thread->second.split_and_retry_oom_injected > 0) {
-          thread->second.split_and_retry_oom_injected--;
-          throw_java_exception(SPLIT_AND_RETRY_OOM_CLASS, "injected SplitAndRetryOOM");
+  /**
+   * Check to see if any threads need to move to BUFN. This should be
+   * called when a task or shuffle thread becomes blocked so that we can 
+   * check to see if one of them needs to becom BUFN or do a split and rollback.
+   */
+  void check_and_update_for_bufn(std::unique_lock<std::mutex> & lock) {
+    // We want to know if all active tasks have at least one thread that
+    // is effectivley blocked or not.  We could change the definitions here,
+    // but for now this sounds like a good starting point.
+    std::set<long> tasks_with_threads;
+    std::set<long> tasks_with_threads_effectively_blocked;
```

Comment:
should this just be a counter? Judging by how it's used for the deadlock busting logic.

Reply:
This is tasks, not threads. So if we have multiple threads on a single task a counter will not work.